x/qgb/types/msgs.go

```diff
@@ -0,0 +1,39 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var _ sdk.Msg = &MsgRegisterEVMAddress{}
+
+func NewMsgRegisterEVMAddress(valAddress, evmAddress string) (*MsgRegisterEVMAddress, error) {
+	msg := &MsgRegisterEVMAddress{
+		ValidatorAddress: valAddress,
+		EvmAddress:       evmAddress,
+	}
+	return msg, msg.ValidateBasic()
+}
+
+// ValidateBasic verifies that the EVM address and val address are of a valid type
+func (msg MsgRegisterEVMAddress) ValidateBasic() error {
+	_, err := sdk.ValAddressFromBech32(msg.ValidatorAddress)
+	if err != nil {
+		return err
+	}
+
+	if !common.IsHexAddress(msg.EvmAddress) {
+		return ErrEVMAddressNotHex
+	}
+
+	return nil
+}
+
+// GetSigner fulfills the sdk.Msg interface. The signer must be the validator address
+func (msg MsgRegisterEVMAddress) GetSigners() []sdk.AccAddress {
+	address, err := sdk.AccAddressFromBech32(msg.ValidatorAddress)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{address}
+}
```

Comment:
[blocking]
if we call this in the tests, it panics since `AccAddressFromBech32` can't handle a `celestiavalop` address. afaiu, this will stop anyone from being able to submit a tx with a MsgRegisterEVMAddress using their valop address. I think this is only working in the tests because we're not actually executing the tx with all of the antehandlers

I think this should work tho
```suggestion
// GetSigner fulfills the sdk.Msg interface. The signer must be the validator address
func (msg MsgRegisterEVMAddress) GetSigners() []sdk.AccAddress {
	valAddr, err := sdk.ValAddressFromBech32(msg.ValidatorAddress)
	if err != nil {
		panic(err)
	}
	return []sdk.AccAddress{sdk.AccAddress(valAddr)}
}
```

Reply:
we also don't have to use this, as I'll defer to the author on how to test these things, but I wanted to make sure that the above section was working as expected, so I quickly wrote

```go
package qgb_test

import (
	"sync"
	"testing"

	"github.com/celestiaorg/celestia-app/app"
	"github.com/celestiaorg/celestia-app/app/encoding"
	"github.com/celestiaorg/celestia-app/test/util/testnode"
	qgbtypes "github.com/celestiaorg/celestia-app/x/qgb/types"
	"github.com/cosmos/cosmos-sdk/crypto/keyring"
	sdk "github.com/cosmos/cosmos-sdk/types"
	"github.com/cosmos/cosmos-sdk/types/errors"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/stretchr/testify/suite"
	abci "github.com/tendermint/tendermint/abci/types"
	tmrand "github.com/tendermint/tendermint/libs/rand"
)

func TestQGBIntegrationSuite(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping qgb integration test in short mode.")
	}
	suite.Run(t, new(QGBIntegrationSuite))
}

type QGBIntegrationSuite struct {
	suite.Suite

	accounts []string
	cctx     testnode.Context
	ecfg     encoding.Config

	mut            sync.Mutex
	accountCounter int
}

func (s *QGBIntegrationSuite) SetupSuite() {
	t := s.T()
	t.Log("setting up integration test suite")

	accounts := make([]string, 10)
	for i := 0; i < len(accounts); i++ {
		accounts[i] = tmrand.Str(9)
	}

	cfg := testnode.DefaultConfig().WithAccounts(accounts)
	cctx, _, _ := testnode.NewNetwork(t, cfg)
	s.accounts = cfg.Accounts
	s.ecfg = encoding.MakeConfig(app.ModuleEncodingRegisters...)
	s.cctx = cctx
}

func (s *QGBIntegrationSuite) unusedAccount() string {
	s.mut.Lock()
	acc := s.accounts[s.accountCounter]
	s.accountCounter++
	s.mut.Unlock()
	return acc
}

func getAddress(account string, kr keyring.Keyring) sdk.AccAddress {
	rec, err := kr.Key(account)
	if err != nil {
		panic(err)
	}
	addr, err := rec.GetAddress()
	if err != nil {
		panic(err)
	}
	return addr
}

func (s *QGBIntegrationSuite) TestQGB() {
	t := s.T()
	type test struct {
		name                  string
		msgFunc               func() (msgs []sdk.Msg, signer string)
		hash                  string
		expectedCheckTxCode   uint32
		expectedDeliverTxCode uint32
	}
	tests := []test{
		{
			name: "edit a qgb validator address",
			msgFunc: func() (msgs []sdk.Msg, signer string) {
				account := "validator"
				valAcc, err := s.cctx.Keyring.Key(account)
				require.NoError(t, err)
				valAddr, err := valAcc.GetAddress()
				require.NoError(t, err)

				rvalAddr := sdk.ValAddress(valAddr)

				msg, err := qgbtypes.NewMsgRegisterEVMAddress(rvalAddr.String(), "0x95222290DD7278Aa3Ddd389Cc1E1d165CC4BAfe5")
				require.NoError(t, err)
				return []sdk.Msg{msg}, account
			},
			expectedCheckTxCode:   abci.CodeTypeOK,
			expectedDeliverTxCode: abci.CodeTypeOK,
		},
		{
			name: "unable to edit a qgb validator address",
			msgFunc: func() (msgs []sdk.Msg, signer string) {
				account := s.unusedAccount()
				valAcc, err := s.cctx.Keyring.Key("validator")
				require.NoError(t, err)
				valAddr, err := valAcc.GetAddress()
				require.NoError(t, err)

				rvalAddr := sdk.ValAddress(valAddr)

				msg, err := qgbtypes.NewMsgRegisterEVMAddress(rvalAddr.String(), "0x95222290DD7278Aa3Ddd389Cc1E1d165CC4BAfe5")
				require.NoError(t, err)
				return []sdk.Msg{msg}, account
			},
			expectedCheckTxCode:   abci.CodeTypeOK,
			expectedDeliverTxCode: errors.ErrInvalidPubKey.ABCICode(),
		},
	}

	// sign and submit the transactions
	for i, tt := range tests {
		msgs, signer := tt.msgFunc()
		res, err := testnode.SignAndBroadcastTx(s.ecfg, s.cctx.Context, signer, msgs...)
		require.NoError(t, err)
		require.NotNil(t, res)
		assert.Equal(t, abci.CodeTypeOK, res.Code, tt.name)
		tests[i].hash = res.TxHash
	}

	require.NoError(s.T(), s.cctx.WaitForNextBlock())

	for _, tt := range tests {
		res, err := testnode.QueryTx(s.cctx.Context, tt.hash, true)
		assert.NoError(t, err)
		require.NotNil(t, res)
		assert.Equal(t, tt.expectedDeliverTxCode, res.TxResult.Code, tt.name)
	}
}

```
feel free to copy paste or adjust if needed

Reply:
Yeah adding an integration test here makes sense. Will try do this. As to whether we use a normal sdk address or a validator operator address for the mapping, I'm still not sure. Need to check how each gets used. It may be simpler to use the normal sdk address that's associated with the validator